### PR TITLE
Add multiple etcd address support.

### LINKED ISSUE #1231: [discuss] feature: support etcd cluster in Apache APISIX.
The etcd cluster is highly available, so Apache APISIX should allow multiple etcd addresses to be configured to connect to the etcd cluster.

How about this new style?

```yaml
etcd:
  host: 
      - "http://127.0.0.1:2379"   # multiple etcd address
      - "http://127.0.0.1:2380"  
  prefix: "/apisix"               # apisix configurations prefix
  timeout: 3                      # 3 seconds
```

BTW, the driver `resty-etcd` has supported this feature.

https://github.com/iresty/lua-resty-etcd/blob/master/t/v2/cluster.t#L52

### PR DESCRIPTION
**Pull request by Akayeshmantha**
Add multiple etcd address support

### Issues resolved

Fix #1231

### CODE REVIEW

**Comment by Akayeshmantha** [comment]
@membphis 

test cases fails in https://github.com/apache/incubator-apisix/blob/master/lua/apisix/core/config_etcd.lua#L350

How the etcd db loaded for the travis encironemnt is it a mock one or a real etcd db running ?

**Comment by membphis** [comment]
> is it a mock one or a real etcd db running ?

It is a real etcd, the same as your local ENV.

**Comment by Akayeshmantha** [comment]
Hi @membphis 

Most of the test cases fails with the below log

`" should not match any line in error.log but matches line "2020/03/15 22:47:57 [error] 17269#17269: *169 [lua] config_etcd.lua:350: failed to fetch data from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/ssl?waitIndex=10100&wait=true&recursive=true err:timeout,  etcd key: /apisix/ssl, context: ngx.timer" (req 0)
# 2020/03/15 22:47:57 [error] 17269#17269: *171 [lua] config_etcd.lua:77: waitdir(): failed to get key from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/global_rules?waitIndex=10100&wait=true&recursive=true err:timeout, context: ngx.timer
# 2020/03/15 22:47:57 [info] 17269#17269: *171 [lua] config_etcd.lua:205: waitdir key: /apisix/global_rules prev_index: 10100, context: ngx.timer
# 2020/03/15 22:47:57 [info] 17269#17269: *171 [lua] config_etcd.lua:206: res: null, context: ngx.timer
# 2020/03/15 22:47:57 [error] 17269#17269: *171 [lua] config_etcd.lua:350: failed to fetch data from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/global_rules?waitIndex=10100&wait=true&recursive=true err:timeout,  etcd key: /apisix/global_rules, context: ngx.timer
# 2020/03/15 22:47:57 [error] 17269#17269: *173 [lua] config_etcd.lua:77: waitdir(): failed to get key from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/services?waitIndex=10100&wait=true&recursive=true err:timeout, context: ngx.timer
# 2020/03/15 22:47:57 [info] 17269#17269: *173 [lua] config_etcd.lua:205: waitdir key: /apisix/services prev_index: 10100, context: ngx.timer
# 2020/03/15 22:47:57 [info] 17269#17269: *173 [lua] config_etcd.lua:206: res: null, context: ngx.timer
# 2020/03/15 22:47:57 [error] 17269#17269: *173 [lua] config_etcd.lua:350: failed to fetch data from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/services?waitIndex=10100&wait=true&recursive=true err:timeout,  etcd key: /apisix/services, context: ngx.timer
# 2020/03/15 22:47:57 [error] 17269#17269: *175 [lua] config_etcd.lua:77: waitdir(): failed to get key from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/proto?waitIndex=10100&wait=true&recursive=true err:timeout, context: ngx.timer
# '
#   at /home/ayesh/Documents/API_ASPIX/incubator-apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 1280.

#   Failed test 'TEST 9: invalid route: wrong count - pattern "[error]" should not match any line in error.log but matches line "2020/03/15 22:47:57 [error] 17269#17269: *171 [lua] config_etcd.lua:77: waitdir(): failed to get key from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/global_rules?waitIndex=10100&wait=true&recursive=true err:timeout, context: ngx.timer" (req 0)
# 2020/03/15 22:47:57 [info] 17269#17269: *171 [lua] config_etcd.lua:205: waitdir key: /apisix/global_rules prev_index: 10100, context: ngx.timer
# 2020/03/15 22:47:57 [info] 17269#17269: *171 [lua] config_etcd.lua:206: res: null, context: ngx.timer
# 2020/03/15 22:47:57 [error] 17269#17269: *171 [lua] config_etcd.lua:350: failed to fetch data from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/global_rules?waitIndex=10100&wait=true&recursive=true err:timeout,  etcd key: /apisix/global_rules, context: ngx.timer
# 2020/03/15 22:47:57 [error] 17269#17269: *173 [lua] config_etcd.lua:77: waitdir(): failed to get key from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/services?waitIndex=10100&wait=true&recursive=true err:timeout, context: ngx.timer
# 2020/03/15 22:47:57 [info] 17269#17269: *173 [lua] config_etcd.lua:205: waitdir key: /apisix/services prev_index: 10100, context: ngx.timer
# 2020/03/15 22:47:57 [info] 17269#17269: *173 [lua] config_etcd.lua:206: res: null, context: ngx.timer
# 2020/03/15 22:47:57 [error] 17269#17269: *173 [lua] config_etcd.lua:350: failed to fetch data from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/services?waitIndex=10100&wait=true&recursive=true err:timeout,  etcd key: /apisix/services, context: ngx.timer
# 2020/03/15 22:47:57 [error] 17269#17269: *175 [lua] config_etcd.lua:77: waitdir(): failed to get key from etcd: fail request to: http://127.0.0.1:2379/v2/keys/apisix/proto?waitIndex=10100&wait=true&recursive=true err:timeout, context: ngx.timer
# 2020/03/15 22:47:57 [info] 17269#17269: *175 [lua] config_etcd.lua:205: waitdir key: /apisix/proto prev_index: 10100, context: ngx.timer
# '
#   at /home/ayesh/Documents/API_ASPIX/incubator-apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 1280.
`

Hm I have no clue is it because now we are defining set of etcd hosts it can't find a leader ?

Some help here would be great

**Comment by membphis** [comment]
I updated the resty-etcd right now. Please rebase your branch with APISIX master branch.

File: conf/config.yaml [hunk 1]
```diff
@@ -98,7 +98,10 @@ nginx_config:                     # config for render the template to genarate n
       - 'unix:'
 
 etcd:
-  host: "http://127.0.0.1:2379"   # etcd address
+  host:
+    - "http://127.0.0.1:2379"     # multiple etcd address
```
Comment (membphis) [inline conf/config.yaml]: That is wrong. By default, there should be only one etcd address.

File: bin/apisix [hunk 2]
```diff
@@ -661,28 +661,32 @@ local function init_etcd(show_output)
     end
 
     local etcd_conf = yaml_conf.etcd
-    local uri = etcd_conf.host .. "/v2/keys" .. (etcd_conf.prefix or "")
-
     local timeout = etcd_conf.timeout or 3
+    local uri
+
+    for _, host in ipairs(yaml_conf.etcd.host) do
```
Comment (membphis) [inline bin/apisix]: this is wrong.

If the user specifies multiple etcd addresses, these etcd instances should belong to one cluster.

When we need to initialize etcd, as long as one of them is successfully initialized, it is considered successful.

File: conf/config.yaml (hunk 1 above)
Comment (Akayeshmantha) [reply conf/config.yaml]: oh sure and multiple http_hosts. I got it wrong

**Comment by membphis** [comment]
you can use `git rebase ...`, it is simpler than creating a new issue.

**Comment by Akayeshmantha** [comment]
@membphis  sure I will do so

File: conf/config.yaml (hunk 1 above)
Comment (Akayeshmantha) [reply conf/config.yaml]: Sure fixed

File: bin/apisix (hunk 2 above)
Comment (Akayeshmantha) [reply bin/apisix]: sure fixed

**Comment by Akayeshmantha** [comment]
@membphis  done opened the previous PR.

File: bin/apisix (hunk 2 above)
Comment (membphis) [reply bin/apisix]: that is the fake code, please take a view.

```lua
-- fake code
for i, host in ipairs(yaml_conf.etcd.host) do
    local is_success = true
    for _, dir_name in ipairs({"/routes", "/upstreams", ...}) do
        local ok = ...
        if not ok then
            is_success = false
            break
        end
    end

    if is_success then
        break
    end
end
```

**Comment by membphis** [comment]
image

**Comment by Akayeshmantha** [comment]
@membphis  but it won't affect the code base know ?. Shall I go on with the other PR. Since it will take more time to redo this I think it's easy to move forward with the other one

wdyt?

**Comment by membphis** [comment]
> Shall I go on with the other PR. 

You can continue working at this PR.

**Comment by Akayeshmantha** [comment]
Hi @membphis  am not quite sure how to fix this I tried in few ways but it has so many conflicts will close this PR and work on the other one 

Thanks

**Pull request closed without merging**